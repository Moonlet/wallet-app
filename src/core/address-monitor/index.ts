import { Notifications } from '../messaging/notifications/notifications';
import { Blockchain } from '../blockchain/types';
import DeviceInfo from 'react-native-device-info';
import { IWalletsState, IAccountState } from '../../redux/wallets/state';
import { CONFIG } from '../../config';

const url = CONFIG.dataApiUrl + '/notifications/register';
const monitoredBlockchains = [Blockchain.ETHEREUM, Blockchain.ZILLIQA];

// const getDeviceToken = (): Promise<string> =>
//     Platform.OS === 'android' ? Notifications.getToken() : getApnsToken();

export const updateAddressMonitorTokens = (wallets: IWalletsState) => {
    const addresses = {};

    Object.keys(wallets || []).forEach(id => {
        wallets[id].accounts.forEach((account: IAccountState) => {
            if (monitoredBlockchains.indexOf(account.blockchain) !== -1) {
                if (!addresses[account.blockchain]) {
                    addresses[account.blockchain] = [];
                }

                addresses[account.blockchain].push(account.address);
            }
        });
    });

    const requestAddresses = Object.keys(addresses).map(blockchain => ({
        blockchain,
        addresses: addresses[blockchain]
    }));

    const deviceId = DeviceInfo.getUniqueId();
    Notifications.getToken().then(token => {
        const request = {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                deviceId,
                tokenType: 'fcm',
                token,
                addresses: requestAddresses
            })
        };

        fetch(url, request);
    });
};
