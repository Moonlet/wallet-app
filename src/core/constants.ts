import { Blockchain } from './blockchain/types';

const BLOCKCHAIN_COINS = {
    [Blockchain.ETHEREUM]: 'ETH',
    [Blockchain.ZILLIQA]: 'ZIL',
    [Blockchain.COSMOS]: 'ATOM',
    [Blockchain.STELLAR]: 'XLM'
};

export { BLOCKCHAIN_COINS };
