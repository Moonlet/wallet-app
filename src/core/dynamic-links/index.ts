import dynamicLinks from '@react-native-firebase/dynamic-links';
import { URL, URLSearchParams } from 'react-native-url-polyfill';
import { captureException as SentryCaptureException } from '@sentry/react-native';
import { IReduxState } from '../../redux/state';
import { NavigationService } from '../../navigation/navigation-service';
import { getChainId } from '../../redux/preferences/selectors';
import { setSelectedBlockchain } from '../../redux/wallets/actions';
import { Blockchain } from '../blockchain/types';
import { getNrPendingTransactions, getSelectedAccount } from '../../redux/wallets/selectors';
import { ITokenState } from '../../redux/wallets/state';
import { store } from '../../redux/config';

export class DynamicLinksService {
    private onLinkListener = null;
    private state: IReduxState = null;

    private handlers = {
        '/:blockchain/:token/validator/:validatorAddress': params => {
            this.navigateValidatorAddress(params);
        },
        '/:blockchain/:token/stake/:validatorAddress': params => {
            this.stakeValidatorAddress(params);
        },
        '/:blockchain/stake': params => {
            this.blockchainStake(params);
        },
        '/stake': params => {
            // TODO: implement later on
            // console.log('stakeDefault', params);
        }
    };

    private navigateValidatorAddress(params: any) {
        const blockchain = String(params?.blockchain)?.toUpperCase() as Blockchain;
        const validatorAddress = params?.validatorAddress;
        const tokenSymbol = String(params?.token)?.toUpperCase();

        if (blockchain && validatorAddress && tokenSymbol) {
            // make sure `blockchain` is selected
            store.dispatch(setSelectedBlockchain(blockchain));

            const chainId = getChainId(this.state, blockchain);

            const selectedAccount = getSelectedAccount(this.state);

            if (!selectedAccount) {
                SentryCaptureException(
                    new Error(
                        `Dynamic Links - no selected account, no wallet, blockchain: ${blockchain}, tokenSymbol: ${tokenSymbol}, validatorAddress: ${validatorAddress}`
                    )
                );
                return;
            }

            const token: ITokenState = selectedAccount.tokens[chainId][tokenSymbol];

            NavigationService.navigate('Validator', {
                blockchain,
                undefined, // validator
                accountIndex: selectedAccount.index,
                token,
                canPerformAction: !getNrPendingTransactions(this.state), // hasPendingTransactions
                options: {
                    validatorAddress,
                    tokenSymbol
                }
            });
        } else {
            SentryCaptureException(
                new Error(`Dynamic Links - navigateValidatorAddress, invalid params: ${params}`)
            );
        }
    }

    private stakeValidatorAddress(params: any) {
        const blockchain = String(params?.blockchain)?.toUpperCase() as Blockchain;
        const validatorAddress = params?.validatorAddress;

        if (blockchain && validatorAddress) {
            // make sure `blockchain` is selected
            store.dispatch(setSelectedBlockchain(blockchain));

            NavigationService.navigate('SmartScreen', {
                context: {
                    screen: 'StakeNow',
                    step: 'SelectStakeAccount',
                    key: 'select-stake-account',
                    params: {
                        validatorId: validatorAddress
                    }
                },
                navigationOptions: {
                    title: 'Stake now'
                },
                newFlow: true
            });
        } else {
            SentryCaptureException(
                new Error(`Dynamic Links - stakeValidatorAddress, invalid params: ${params}`)
            );
        }
    }

    private blockchainStake(params: any) {
        const blockchain = String(params?.blockchain)?.toUpperCase() as Blockchain;

        if (blockchain) {
            // make sure `blockchain` is selected
            store.dispatch(setSelectedBlockchain(blockchain));

            NavigationService.navigate('SmartScreen', {
                context: {
                    screen: 'StakeNow',
                    step: 'StakeSelectValidator',
                    key: 'stake-now-select-validator'
                },
                navigationOptions: {
                    title: 'Stake now'
                },
                newFlow: true
            });
        } else {
            SentryCaptureException(
                new Error(`Dynamic Links - blockchainStake, invalid params: ${params}`)
            );
        }
    }

    public async configure(state: IReduxState) {
        this.state = state;

        const dl = dynamicLinks();

        this.onLinkListener = dl.onLink(link => {
            this.handleDynamicLink(link);
        });

        // application is in a background state or has fully quit
        dl.getInitialLink().then(link => {
            this.handleDynamicLink(link);
        });
    }

    // app is in the foreground state
    public async handleDynamicLink(link: any) {
        if (link?.url) {
            this.processUrl(link.url, this.handlers, {
                protocol: ['https:'],
                host: [
                    'moonlet.io',
                    'moonlet.xyz',
                    'moonletwallet.com',
                    'moonlet.app',
                    'moonlet.link'
                ]
            });
        } else {
            // Invalid link url, link:${link}`
        }
    }

    public removeListeners() {
        typeof this.onLinkListener === 'function' && this.onLinkListener();
    }

    private pathPatternToRegexp(pathPattern: string) {
        let exp = '^';

        const pathParts = pathPattern.split('/').filter(Boolean);
        const capturingGroups = [];
        for (const part of pathParts) {
            exp += '\\/';
            if (part.indexOf(':') === 0) {
                exp += '([^\\/]*)';
                capturingGroups.push(part.replace(/^:/, ''));
            } else {
                exp += part;
            }
        }

        exp += '$';
        return {
            regexp: new RegExp(exp),
            capturingGroups
        };
    }

    private parsePath(path: string, pathPatternString: string) {
        const pathPattern = this.pathPatternToRegexp(pathPatternString);
        const match = path.match(pathPattern.regexp);

        if (match) {
            const params = {};
            for (let i = 0; i < pathPattern.capturingGroups.length; i++) {
                params[pathPattern.capturingGroups[i]] = match[i + 1];
            }
            return params;
        }

        return null;
    }

    private processPath(path: string, handlers: any, urlParams = {}) {
        const pathsPatterns = Object.keys(handlers);
        // console.log(path, handlers, urlParams);
        for (const pathPattern of pathsPatterns) {
            if (typeof handlers[pathPattern] === 'function') {
                const params = this.parsePath(path, pathPattern);
                if (params) {
                    return handlers[pathPattern]({
                        ...urlParams,
                        ...params
                    });
                }
            }
        }
    }

    private paramsToObject(entries: any) {
        const result = {};
        for (const [key, value] of entries) {
            // each 'entry' is a [key, value] tupple
            result[key] = value;
        }
        return result;
    }

    private processUrl(url: string, pathHandlers: any = this.handlers, whitelist = {}) {
        const u = new URL(url);

        for (const key of Object.keys(whitelist)) {
            if (u && typeof u[key] === 'string' && whitelist[key].indexOf(u[key]) < 0) {
                SentryCaptureException(
                    new Error(
                        JSON.stringify(
                            `whitelist not matched ${key}=${u[key]} not in whitelist: ${
                                typeof whitelist[key] === 'string'
                                    ? whitelist[key]
                                    : whitelist[key].join(', ')
                            }`
                        )
                    )
                );
                return null;
            }
        }

        let path = u?.pathname;

        if (path && String(path).startsWith('/links')) path = path.replace('/links', '');

        const searchParams = new URLSearchParams(url);

        return this.processPath(path, pathHandlers, this.paramsToObject(searchParams.entries()));
    }
}

export const DynamicLinks = new DynamicLinksService();
