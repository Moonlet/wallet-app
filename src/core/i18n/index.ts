export * from './translation/translate-component';
export * from './translation/translate';
