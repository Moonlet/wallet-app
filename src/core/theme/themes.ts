export enum Theme {
    dark = 'dark',
    light = 'light'
}
