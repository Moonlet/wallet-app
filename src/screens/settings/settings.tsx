import React from 'react';
import { ScrollView, View, Switch, TouchableOpacity, Platform, Clipboard } from 'react-native';
import { INavigationProps } from '../../navigation/with-navigation-params';
import { Text, Button } from '../../library';
import { IReduxState } from '../../redux/state';
import { toggleTouchID } from '../../redux/preferences/actions';
import stylesProvider from './styles';
import { withTheme, IThemeProps } from '../../core/theme/with-theme';
import { Icon } from '../../components/icon';
import { smartConnect } from '../../core/utils/smart-connect';
import { connect } from 'react-redux';
import DeviceInfo from 'react-native-device-info';
import { HeaderIcon } from '../../components/header-icon/header-icon';
import { translate } from '../../core/i18n';
import { biometricAuth, BiometryType } from '../../core/biometric-auth/biometric-auth';
import { PasswordModal } from '../../components/password-modal/password-modal';
import { WalletConnectWeb } from '../../core/wallet-connect/wallet-connect-web';
import { Dialog } from '../../components/dialog/dialog';
import { changePIN } from '../../redux/wallets/actions';
import { isFeatureActive, RemoteFeature } from '../../core/utils/remote-feature-config';
import { DebugModal } from '../../components/debug-modal/debug-modal';

export interface IState {
    isTouchIDSupported: boolean;
    biometryType: BiometryType;
    changePIN: boolean;
}

export interface IReduxProps {
    currency: string;
    touchID: boolean;
    toggleTouchID: typeof toggleTouchID;
    mock: () => void;
    changePIN: typeof changePIN;
}

export const mockFunction = () => {
    return { type: 'dummy' };
};

const mapStateToProps = (state: IReduxState) => ({
    touchID: state.preferences.touchID,
    currency: state.preferences.currency
});

const mapDispatchToProps = {
    mock: mockFunction,
    toggleTouchID,
    changePIN
};

const navigationOptions = () => ({
    title: translate('App.labels.settings'),
    headerLeft: <HeaderIcon />
});

export class SettingsScreenComponent extends React.Component<
    INavigationProps & IReduxProps & IThemeProps<ReturnType<typeof stylesProvider>>,
    IState
> {
    public static navigationOptions = navigationOptions;
    public passwordModal = null;
    debugModal: any;

    constructor(
        props: INavigationProps & IReduxProps & IThemeProps<ReturnType<typeof stylesProvider>>
    ) {
        super(props);

        this.state = {
            isTouchIDSupported: false,
            biometryType: undefined,
            changePIN: false
        };

        biometricAuth
            .isSupported()
            .then(biometryType => {
                // biometryType on Android is of type boolean
                // biometryType on iOS is of type 'FaceID' | 'TouchID'

                if (Platform.OS === 'ios') {
                    this.setState({ biometryType });
                }
                this.setState({ isTouchIDSupported: true });
            })
            .catch(() => {
                // Failure code if the user's device does not have touchID or faceID enabled
                this.setState({ isTouchIDSupported: false });
            });
    }

    public reportIssueTouch = () => {
        // report an issue
        this.props.mock();
    };

    public renderSecuritySection = () => {
        const { styles, navigation, theme } = this.props;

        return (
            <View>
                <Text style={styles.textHeader}>
                    {translate('App.labels.security').toUpperCase()}
                </Text>

                {this.state.isTouchIDSupported && (
                    <View>
                        <View style={styles.rowContainer}>
                            <Text style={styles.textRow}>
                                {Platform.OS === 'ios'
                                    ? translate(`BiometryType.${this.state.biometryType}`)
                                    : translate('BiometryType.FingerprintLogin')}
                            </Text>
                            <Switch
                                onValueChange={() =>
                                    this.passwordModal
                                        .requestPassword()
                                        .then(() => this.props.toggleTouchID())
                                }
                                value={this.props.touchID}
                                trackColor={{
                                    true: this.props.theme.colors.cardBackground,
                                    false: this.props.theme.colors.cardBackground
                                }}
                                thumbColor={
                                    this.props.touchID
                                        ? theme.colors.accent
                                        : theme.colors.textTertiary
                                }
                            />
                        </View>

                        <View style={styles.divider} />
                    </View>
                )}

                <TouchableOpacity
                    style={styles.rowContainer}
                    onPress={() => navigation.navigate('Wallets')}
                >
                    <Text style={styles.textRow}>{translate('Settings.manageWallet')}</Text>
                    <View style={styles.rightContainer}>
                        <Icon name="chevron-right" size={16} style={styles.icon} />
                    </View>
                </TouchableOpacity>

                <View style={styles.divider} />

                <TouchableOpacity
                    style={styles.rowContainer}
                    onPress={() => navigation.navigate('BackupWallet')}
                >
                    <Text style={styles.textRow}>{translate('Settings.backupWallet')}</Text>
                    <View style={styles.rightContainer}>
                        <Icon name="chevron-right" size={16} style={styles.icon} />
                    </View>
                </TouchableOpacity>

                <View style={styles.divider} />

                <TouchableOpacity
                    style={styles.rowContainer}
                    onPress={() =>
                        this.setState({ changePIN: true }, () => {
                            this.passwordModal
                                .requestPassword()
                                .then((pass: { newPassword: string; oldPassword: string }) => {
                                    this.props.changePIN(pass.newPassword, pass.oldPassword);

                                    // disable changePIN if you want to reattempt to change the PIN code
                                    this.setState({ changePIN: false }, () =>
                                        Dialog.info(
                                            translate('App.labels.success'),
                                            translate('Settings.successChangePin')
                                        )
                                    );
                                });
                        })
                    }
                >
                    <Text style={styles.textRow}>{translate('Settings.changePin')}</Text>
                    <Icon name="chevron-right" size={16} style={styles.icon} />
                </TouchableOpacity>

                <View style={styles.divider} />
            </View>
        );
    };

    public renderSetupSection = () => {
        const { styles, navigation } = this.props;

        return (
            <View>
                <Text style={styles.textHeader}>{translate('App.labels.setup').toUpperCase()}</Text>

                <TouchableOpacity
                    style={styles.rowContainer}
                    onPress={() => navigation.navigate('SetCurrency')}
                >
                    <Text style={styles.textRow}>{translate('Settings.defaultCurrency')}</Text>
                    <View style={styles.rightContainer}>
                        <Text style={styles.rightValue}>{this.props.currency}</Text>
                        <Icon name="chevron-right" size={16} style={styles.icon} />
                    </View>
                </TouchableOpacity>

                <View style={styles.divider} />

                <TouchableOpacity
                    style={styles.rowContainer}
                    onPress={() => navigation.navigate('BlockchainPortfolio')}
                >
                    <Text style={styles.textRow}>{translate('Settings.blockchainPortfolio')}</Text>
                    <Icon name="chevron-right" size={16} style={styles.icon} />
                </TouchableOpacity>

                <View style={styles.divider} />
            </View>
        );
    };

    public renderSupportSection = () => {
        const { styles } = this.props;

        return (
            <View>
                <Text style={styles.textHeader}>
                    {translate('App.labels.support').toUpperCase()}
                </Text>

                <TouchableOpacity
                    testID={'report-issue'}
                    style={styles.rowContainer}
                    onPress={this.reportIssueTouch}
                >
                    <Text style={styles.textRow}>{translate('Settings.reportIssue')}</Text>
                    <View style={styles.rightContainer}>
                        <Icon name="chevron-right" size={16} style={styles.icon} />
                    </View>
                </TouchableOpacity>

                <View style={styles.divider} />
            </View>
        );
    };

    public renderToolsSection = () => {
        const { styles, navigation } = this.props;

        return (
            <View>
                <Text style={styles.textHeader}>{translate('App.labels.tools').toUpperCase()}</Text>

                <TouchableOpacity
                    style={styles.rowContainer}
                    onPress={() => navigation.navigate('NetworkOptions')}
                >
                    <Text style={styles.textRow}>{translate('Settings.mainnetTestnet')}</Text>
                    <View style={styles.rightContainer}>
                        <Icon name="chevron-right" size={16} style={styles.icon} />
                    </View>
                </TouchableOpacity>

                <View style={styles.divider} />
            </View>
        );
    };

    public renderAboutSection = () => {
        const { styles, navigation } = this.props;

        return (
            <View>
                <Text style={styles.textHeader}>{translate('App.labels.about').toUpperCase()}</Text>

                <TouchableOpacity
                    style={styles.rowContainer}
                    onPress={() => navigation.navigate('TermsConditions')}
                >
                    <Text style={styles.textRow}>{translate('App.labels.tc')}</Text>
                    <View style={styles.rightContainer}>
                        <Icon name="chevron-right" size={16} style={styles.icon} />
                    </View>
                </TouchableOpacity>

                <View style={styles.divider} />

                <TouchableOpacity
                    style={styles.rowContainer}
                    onPress={() => navigation.navigate('PrivacyPolicy')}
                >
                    <Text style={styles.textRow}>{translate('Settings.privacyPolicy')}</Text>
                    <View style={styles.rightContainer}>
                        <Icon name="chevron-right" size={16} style={styles.icon} />
                    </View>
                </TouchableOpacity>

                <View style={styles.divider} />

                <View style={styles.rowContainer}>
                    <Text style={styles.textRow}>{translate('Settings.appVersion')}</Text>
                    <View style={styles.rightContainer}>
                        <Text style={styles.rightValue}>{DeviceInfo.getVersion()}</Text>
                    </View>
                </View>

                <View style={styles.divider} />

                <TouchableOpacity
                    style={styles.colContainer}
                    onPress={() => {
                        Clipboard.setString(DeviceInfo.getUniqueId());
                        Dialog.info(translate('App.labels.success'), translate('Settings.copied'));
                    }}
                >
                    <Text style={[styles.textRow, styles.textRowMargin]}>
                        {translate('Settings.deviceId')}
                    </Text>
                    <Text style={styles.rightValue}>{DeviceInfo.getUniqueId()}</Text>
                </TouchableOpacity>

                <View style={styles.divider} />
            </View>
        );
    };

    public render() {
        const { styles } = this.props;

        return (
            <View style={styles.container}>
                <ScrollView
                    contentContainerStyle={{ flexGrow: 1 }}
                    showsVerticalScrollIndicator={false}
                >
                    {Platform.OS !== 'web' && this.renderSecuritySection()}
                    {Platform.OS !== 'web' && this.renderSetupSection()}
                    {this.renderSupportSection()}
                    {Platform.OS !== 'web' && this.renderToolsSection()}
                    {this.renderAboutSection()}
                    {isFeatureActive(RemoteFeature.DEV_TOOLS) && (
                        <TouchableOpacity
                            style={styles.rowContainer}
                            onPress={() => {
                                this.debugModal.showDebug();
                            }}
                        >
                            <Text style={styles.textRow}>Debug Info</Text>
                            <View style={styles.rightContainer}>
                                <Icon name="chevron-right" size={16} style={styles.icon} />
                            </View>
                        </TouchableOpacity>
                    )}
                    {Platform.OS === 'web' && (
                        <Button
                            style={styles.button}
                            onPress={() => {
                                WalletConnectWeb.disconnect();
                                location.reload();
                            }}
                        >
                            {translate('App.labels.disconnect')}
                        </Button>
                    )}
                </ScrollView>

                <PasswordModal
                    obRef={ref => (this.passwordModal = ref)}
                    changePIN={this.state.changePIN}
                />
                <DebugModal obRef={ref => (this.debugModal = ref)} />
            </View>
        );
    }
}

export const SettingsScreen = smartConnect(SettingsScreenComponent, [
    connect(mapStateToProps, mapDispatchToProps),
    withTheme(stylesProvider)
]);
