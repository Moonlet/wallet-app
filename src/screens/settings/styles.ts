import { StyleSheet, Platform } from 'react-native';
import { ITheme } from '../../core/theme/itheme';
import { BASE_DIMENSION, normalize } from '../../styles/dimensions';

export default (theme: ITheme) =>
    StyleSheet.create({
        container: {
            flex: 1,
            flexDirection: 'column',
            backgroundColor: theme.colors.appBackground,
            height: Platform.OS === 'web' ? 'calc(100vh - 122px)' : 'auto'
        },
        scrollContainer: {
            flexGrow: 1,
            paddingHorizontal: BASE_DIMENSION * 2
        },
        rowContainer: {
            flexDirection: 'row',
            paddingVertical: BASE_DIMENSION * 2,
            alignItems: 'center'
        },
        colContainer: {
            flexDirection: 'column',
            paddingVertical: BASE_DIMENSION * 2
        },
        textHeader: {
            color: theme.colors.textSecondary,
            fontSize: normalize(13),
            lineHeight: normalize(18),
            marginTop: BASE_DIMENSION * 2
        },
        textRow: {
            flex: 1,
            fontSize: normalize(17),
            lineHeight: normalize(21),
            letterSpacing: 0.38,
            color: theme.colors.text
        },
        textRowMargin: {
            marginBottom: BASE_DIMENSION
        },
        rightValue: {
            fontSize: normalize(15),
            lineHeight: normalize(20),
            color: theme.colors.textTertiary,
            paddingRight: BASE_DIMENSION
        },
        divider: {
            width: '100%',
            height: 1,
            backgroundColor: theme.colors.settingsDivider
        },
        rightContainer: {
            flex: 1,
            flexDirection: 'row',
            justifyContent: 'flex-end',
            alignItems: 'center'
        },
        icon: {
            color: theme.colors.accent
        },
        button: {
            margin: BASE_DIMENSION * 3
        }
    });
