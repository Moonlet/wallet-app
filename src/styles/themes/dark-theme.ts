import { ITheme } from '../../core/theme/itheme';
import { COLORS } from '../colors';

export const darkTheme: ITheme = {
    dark: true,
    shadowGradient: ['#00000000', '#000000CC'],

    fontSize: {
        small: 12,
        regular: 16,
        large: 34
    },

    colors: {
        primary: COLORS.GRAY,
        accent: COLORS.AQUA,
        text: COLORS.WHITE,
        textDarker: COLORS.LIGHT_GRAY,
        positive: COLORS.HOT_GREEN,
        negative: COLORS.HOT_RED,

        cardBackground: COLORS.DARK_GRAY,
        appBackground: COLORS.COD_GRAY,
        settingsDivider: COLORS.BLACK
    }
};
