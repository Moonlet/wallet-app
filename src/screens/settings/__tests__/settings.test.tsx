import 'react-native';
import React from 'react';
import { SettingsScreenComponent, IProps, IReduxProps, mockFunction } from '../settings';
import { shallow } from 'enzyme';
import styleProvider from '../styles';
import { darkTheme } from '../../../styles/themes/dark-theme';
import { loadTranslations } from '../../../core/i18n';

const props: IProps & IReduxProps = {
    // @ts-ignore
    navigation: {
        navigate: jest.fn()
    },
    styles: styleProvider(darkTheme),
    theme: darkTheme,
    mock: jest.fn(),
    pinLogin: true,
    setPinLogin: jest.fn()
};
beforeEach(() => {
    // @ts-ignore
    props.mock.mockClear();
});

export default describe('Wallet screen', () => {
    beforeAll(async () => {
        await loadTranslations('en');
    });

    it('renders correctly', () => {
        const wrapper = shallow(<SettingsScreenComponent {...props} />);
        expect(wrapper.debug()).toMatchSnapshot();
    });
    it('Pin login switch tapped', () => {
        const wrapper = shallow(<SettingsScreenComponent {...props} />);
        wrapper.find('[testID="pin-login"]').simulate('valueChange');
        expect(props.setPinLogin).toHaveBeenCalledTimes(1);
    });
    it('Touch id switch tapped', () => {
        const wrapper = shallow(<SettingsScreenComponent {...props} />);
        wrapper.find('[testID="touch-id"]').simulate('valueChange');
        expect(props.mock).toHaveBeenCalled();
    });
    it('Backup Wallet', () => {
        const wrapper = shallow(<SettingsScreenComponent {...props} />);
        wrapper.find('[testID="backup-wallet"]').simulate('Press');
        expect(props.mock).toHaveBeenCalledTimes(1);
    });
    it('Blockchain portfolio', () => {
        const wrapper = shallow(<SettingsScreenComponent {...props} />);
        wrapper.find('[testID="blockchain-portfolio"]').simulate('Press');
        expect(props.mock).toHaveBeenCalledTimes(1);
    });
    it('Report issue', () => {
        const wrapper = shallow(<SettingsScreenComponent {...props} />);
        wrapper.find('[testID="report-issue"]').simulate('Press');
        expect(props.mock).toHaveBeenCalledTimes(1);
    });
});
