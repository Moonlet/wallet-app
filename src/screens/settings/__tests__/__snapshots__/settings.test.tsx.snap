// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wallet screen renders correctly 1`] = `
"<View style={{...}}>
  <ScrollViewMock contentContainerStyle={{...}} showsVerticalScrollIndicator={false}>
    <View>
      <withTheme(TextComponent) style={{...}}>
        SECURITY
      </withTheme(TextComponent)>
      <TouchableOpacity style={{...}} onPress={[Function: onPress]} activeOpacity={0.2}>
        <withTheme(TextComponent) style={{...}}>
          Manage wallet
        </withTheme(TextComponent)>
        <View style={{...}}>
          <Icon name=\\"chevron-right\\" size={16} style={{...}} allowFontScaling={false} />
        </View>
      </TouchableOpacity>
      <View style={{...}} />
      <TouchableOpacity style={{...}} onPress={[Function: onPress]} activeOpacity={0.2}>
        <withTheme(TextComponent) style={{...}}>
          Backup wallet
        </withTheme(TextComponent)>
        <View style={{...}}>
          <Icon name=\\"chevron-right\\" size={16} style={{...}} allowFontScaling={false} />
        </View>
      </TouchableOpacity>
      <View style={{...}} />
      <TouchableOpacity style={{...}} onPress={[Function: onPress]} activeOpacity={0.2}>
        <withTheme(TextComponent) style={{...}}>
          Change PIN
        </withTheme(TextComponent)>
        <Icon name=\\"chevron-right\\" size={16} style={{...}} allowFontScaling={false} />
      </TouchableOpacity>
      <View style={{...}} />
    </View>
    <View>
      <withTheme(TextComponent) style={{...}}>
        SETUP
      </withTheme(TextComponent)>
      <TouchableOpacity style={{...}} onPress={[Function: onPress]} activeOpacity={0.2}>
        <withTheme(TextComponent) style={{...}}>
          Default currency
        </withTheme(TextComponent)>
        <View style={{...}}>
          <withTheme(TextComponent) style={{...}} />
          <Icon name=\\"chevron-right\\" size={16} style={{...}} allowFontScaling={false} />
        </View>
      </TouchableOpacity>
      <View style={{...}} />
      <TouchableOpacity style={{...}} onPress={[Function: onPress]} activeOpacity={0.2}>
        <withTheme(TextComponent) style={{...}}>
          Blockchain portfolio
        </withTheme(TextComponent)>
        <Icon name=\\"chevron-right\\" size={16} style={{...}} allowFontScaling={false} />
      </TouchableOpacity>
      <View style={{...}} />
    </View>
    <View>
      <withTheme(TextComponent) style={{...}}>
        SUPPORT
      </withTheme(TextComponent)>
      <TouchableOpacity testID=\\"report-issue\\" style={{...}} onPress={[Function]} activeOpacity={0.2}>
        <withTheme(TextComponent) style={{...}}>
          Report issue
        </withTheme(TextComponent)>
        <View style={{...}}>
          <Icon name=\\"chevron-right\\" size={16} style={{...}} allowFontScaling={false} />
        </View>
      </TouchableOpacity>
      <View style={{...}} />
    </View>
    <View>
      <withTheme(TextComponent) style={{...}}>
        TOOLS
      </withTheme(TextComponent)>
      <TouchableOpacity style={{...}} onPress={[Function: onPress]} activeOpacity={0.2}>
        <withTheme(TextComponent) style={{...}}>
          Mainnet/Testnet
        </withTheme(TextComponent)>
        <View style={{...}}>
          <Icon name=\\"chevron-right\\" size={16} style={{...}} allowFontScaling={false} />
        </View>
      </TouchableOpacity>
      <View style={{...}} />
    </View>
    <View>
      <withTheme(TextComponent) style={{...}}>
        ABOUT
      </withTheme(TextComponent)>
      <TouchableOpacity style={{...}} onPress={[Function: onPress]} activeOpacity={0.2}>
        <withTheme(TextComponent) style={{...}}>
          Terms &amp; conditions
        </withTheme(TextComponent)>
        <View style={{...}}>
          <Icon name=\\"chevron-right\\" size={16} style={{...}} allowFontScaling={false} />
        </View>
      </TouchableOpacity>
      <View style={{...}} />
      <TouchableOpacity style={{...}} onPress={[Function: onPress]} activeOpacity={0.2}>
        <withTheme(TextComponent) style={{...}}>
          Privacy policy
        </withTheme(TextComponent)>
        <View style={{...}}>
          <Icon name=\\"chevron-right\\" size={16} style={{...}} allowFontScaling={false} />
        </View>
      </TouchableOpacity>
      <View style={{...}} />
      <View style={{...}}>
        <withTheme(TextComponent) style={{...}}>
          Application version
        </withTheme(TextComponent)>
        <View style={{...}}>
          <withTheme(TextComponent) style={{...}} />
        </View>
      </View>
      <View style={{...}} />
      <TouchableOpacity style={{...}} onPress={[Function: onPress]} activeOpacity={0.2}>
        <withTheme(TextComponent) style={{...}}>
          Device ID
        </withTheme(TextComponent)>
        <withTheme(TextComponent) style={{...}} />
      </TouchableOpacity>
      <View style={{...}} />
    </View>
    <TouchableOpacity style={{...}} onPress={[Function: onPress]} activeOpacity={0.2}>
      <withTheme(TextComponent) style={{...}}>
        Debug Info
      </withTheme(TextComponent)>
      <View style={{...}}>
        <Icon name=\\"chevron-right\\" size={16} style={{...}} allowFontScaling={false} />
      </View>
    </TouchableOpacity>
  </ScrollViewMock>
  <withTheme(PasswordModalComponent) obRef={[Function: obRef]} changePIN={false} />
  <withTheme(DebugModalComponent) obRef={[Function: obRef]} />
</View>"
`;
